app/invocation/invocation_tabs.tsx

```diff
@@ -1,39 +1,64 @@
 import React from "react";
+import { ToolName } from "./invocation_model";
 
-interface Props {
+export type InvocationTabsProps = TabsContext;
+
+/**
+ * TabsContext contains common props that determine how tabs are rendered
+ * and which tab is the default tab.
+ */
+export type TabsContext = {
   hash: string;
+  denseMode: boolean;
+  tool: ToolName;
+};
+
+export type InvocationTabHash =
+  | "#" // "ALL" tab
+  | "#targets"
+  | "#log"
+  | "#details"
+  | "#artifacts"
+  | "#timing"
+  | "#cache"
+  | "#raw"
+  | "#execution"; // "EXECUTION" tab (currently not rendered).
+
+/** Returns the `TabHash` of the tab that is active when the URL hash is empty. */
+export function getDefaultTab({ tool, denseMode }: Omit<TabsContext, "hash">): InvocationTabHash {
+  if (!denseMode) return "#";
+
+  return tool === "ci_runner" ? "#log" : "#targets";
 }
 
-export default class InvocationTabsComponent extends React.Component {
-  props: Props;
+/** Returns whether the given tab is selected. */
+export function isTabSelected(tabHash: InvocationTabHash, { tool, hash, denseMode }: TabsContext) {
+  return tabHash === (hash || getDefaultTab({ tool, denseMode }));
+}
+
+export default class InvocationTabsComponent extends React.Component<InvocationTabsProps> {
+  private renderTab(tab: InvocationTabHash, { label }: { label: string }) {
+    const { hash, denseMode, tool } = this.props;
```

Comment:
I know I've said this before, but I feel like this destructuring is so much more confusing than just using this.props.hash, this.props.denseMode and this.props.tool throughout. Feel free to ignore.

Reply:
I can see how it might be confusing -- the syntax does look a little odd, and if both props and state are destructured then some of the meaning behind which are props and which are state gets lost.

I personally think the destructuring is cleaner, but I'll avoid it for now just for the sake of consistency.